Value macro not able to handle `Option` in Dict

In GitLab by @seiwill on Feb 1, 2023, 17:41

Deriving `Value` on `signature = dict` `Struct`s that have `Option` values fails to compile.

```rust
use zbus::zvariant::{DeserializeDict, Type, Value};

#[derive(Debug, Clone, PartialEq, Default, Type, DeserializeDict, Value)]
#[zvariant(signature = "dict")]
pub struct ConnectionSettings {
    pub val_bool: bool,
    pub opt_string: Option<String>,
}

fn main() {
    println!("{:?}", ConnectionSettings::default());
}
```

Gives error:

```rust
error[E0277]: the trait bound `std::option::Option<std::string::String>: From<zbus::zvariant::Value<'_>>` is not satisfied
   --> src/main.rs:44:67
    |
44  | #[derive(Debug, Clone, PartialEq, Default, Type, DeserializeDict, Value)]
    |                                                                   ^^^^^ the trait `From<zbus::zvariant::Value<'_>>` is not implemented for `std::option::Option<std::string::String>`
    |
    = help: the following other types implement trait `From<T>`:
              <std::option::Option<&'a T> as From<&'a std::option::Option<T>>>
              <std::option::Option<&'a mut T> as From<&'a mut std::option::Option<T>>>
              <std::option::Option<&'a tracing_core::span::Id> as From<&'a tracing::span::EnteredSpan>>
              <std::option::Option<&'a tracing_core::span::Id> as From<&'a tracing::span::Span>>
              <std::option::Option<&'a tracing_core::span::Id> as From<&'a tracing_core::span::Current>>
              <std::option::Option<&'static tracing_core::metadata::Metadata<'static>> as From<&'a tracing_core::span::Current>>
              <std::option::Option<T> as From<Optional<T>>>
              <std::option::Option<T> as From<T>>
            and 7 others
    = note: required because of the requirements on the impl of `Into<std::option::Option<std::string::String>>` for `zbus::zvariant::Value<'_>`
    = note: required because of the requirements on the impl of `TryFrom<zbus::zvariant::Value<'_>>` for `std::option::Option<std::string::String>`
note: required by a bound in `zbus::zvariant::Value::<'a>::downcast`
   --> /home/sw/.cargo/registry/src/github.com-1ecc6299db9ec823/zvariant-3.10.0/src/value.rs:310:12
    |
310 |         T: TryFrom<Value<'a>>,
    |            ^^^^^^^^^^^^^^^^^^ required by this bound in `zbus::zvariant::Value::<'a>::downcast`
    = note: this error originates in the derive macro `Value` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
```

--- a/zvariant_derive/src/value.rs
+++ b/zvariant_derive/src/value.rs
@@ -4,6 +4,7 @@ use syn::{
     spanned::Spanned, Attribute, Data, DataEnum, DeriveInput, Error, Fields, Generics, Ident,
     Lifetime, LifetimeParam,
 };
+use zvariant_utils::macros;
 
 use crate::utils::*;
 
@@ -123,32 +124,70 @@ fn impl_struct(
                 .map(|field| field.ident.to_token_stream())
                 .collect();
             let (from_value_impl, into_value_impl) = match signature {
-                Some(signature) if signature == "a{sv}" => (
+                Some(signature) if signature == "a{sv}" => {
                     // User wants the type to be encoded as a dict.
                     // FIXME: Not the most efficient implementation.
-                    quote! {
-                        let mut fields = <::std::collections::HashMap::<::std::string::String, #zv::Value>>::try_from(value)?;
+                    let (fields_init, entries_init): (TokenStream, TokenStream) = fields
+                        .iter()
+                        .map(|field| {
+                            let field_name = field.ident.to_token_stream();
+                            let convert = if macros::ty_is_option(&field.ty) {
+                                quote! {
+                                    .map(#zv::Value::downcast)
+                                    .transpose()?
+                                }
+                            } else {
+                                quote! {
+                                    .ok_or_else(|| #zv::Error::IncorrectType)?
+                                    .downcast()?
+                                }
+                            };
 
-                        ::std::result::Result::Ok(Self {
-                            #(
-                                #field_names:
-                                    fields
-                                        .remove(stringify!(#field_names))
-                                        .ok_or_else(|| #zv::Error::IncorrectType)?
-                                        .downcast()?
-                            ),*
+                            let fields_init = quote! {
+                                #field_name: fields
+                                    .remove(stringify!(#field_name))
+                                    #convert,
+                            };
+                            let entries_init = if macros::ty_is_option(&field.ty) {
+                                quote! {
+                                    if let Some(v) = s.#field_name {
+                                        fields.insert(
+                                            stringify!(#field_name),
+                                            #zv::Value::from(v),
+                                        );
+                                    }
+                                }
+                            } else {
+                                quote! {
+                                    fields.insert(
+                                        stringify!(#field_name),
+                                        #zv::Value::from(s.#field_name),
+                                    );
+                                }
+                            };
+
+                            (fields_init, entries_init)
                         })
-                    },
-                    quote! {
-                        let mut fields = ::std::collections::HashMap::new();
-                        #(
-                            fields.insert(stringify!(#field_names), #zv::Value::from(s.#field_names));
-                        )*
+                        .unzip();
 
-                        <#value_type>::#into_value_method(#zv::Value::from(fields))
-                            #into_value_error_transform
-                    },
-                ),
+                    (
+                        quote! {
+                            let mut fields = <::std::collections::HashMap::<
+                                ::std::string::String,
+                                #zv::Value,
+                            >>::try_from(value)?;
+
+                            ::std::result::Result::Ok(Self { #fields_init })
+                        },
+                        quote! {
+                            let mut fields = ::std::collections::HashMap::new();
+                            #entries_init
+
+                            <#value_type>::#into_value_method(#zv::Value::from(fields))
+                                #into_value_error_transform
+                        },
+                    )
+                }
                 Some(_) | None => (
                     quote! {
                         let mut fields = #zv::Structure::try_from(value)?.into_fields();

--- a/zvariant_derive/tests/tests.rs
+++ b/zvariant_derive/tests/tests.rs
@@ -82,3 +82,20 @@ fn derive_dict() {
 
     assert_eq!(Test::SIGNATURE, "a{sv}")
 }
+
+#[test]
+#[ignore]
+fn issues_311() {
+    // Issue 311: Value macro not able to handle Option in Dict.
+    //
+    // org.freedesktop.ModemManager1.Modem.Signal props are a dict with optional values depending on
+    // the property you read.
+    #[derive(Debug, Type, DeserializeDict, OwnedValue, Value)]
+    #[zbus(signature = "dict")]
+    pub struct SignalInfo {
+        pub rssi: Option<i32>,
+        pub ecio: Option<i32>,
+        pub io: Option<i32>,
+        pub sinr: Option<i32>,
+    }
+}

